### PR TITLE
24830 - extract PIN allocation to database

### PR DESCRIPTION
**Pull request by jon-shipley**
Pin changes to HPA feature paths:

- create an auxiliary table for pins
- add a stored procedure to create checks that allocates an unused pin from the aux table
- only requires flat time when allocating a pin - removes network round trips when a sizeable amount of the pin-space has been already used

### CODE REVIEW

File: admin/data/sql/migrations/144.do.create-pin-table.sql [hunk 1]
```diff
@@ -0,0 +1,14 @@
+-- create a pin table
+DROP TABLE IF EXISTS [mtc_admin].[pin];
+create table [mtc_admin].[pin] (
+  [id] [int] IDENTITY(1,1) NOT NULL,
+  [createdAt] [datetimeoffset](3) NOT NULL
+    CONSTRAINT[DF_pin_created_at] DEFAULT CURRENT_TIMESTAMP,
```
Comment (GuyHarwood) [inline admin/data/sql/migrations/144.do.create-pin-table.sql]: we should use `GETUTCDATE()` for consistency

File: admin/data/sql/migrations/147.do.create-check-table-type.sql [hunk 2]
```diff
@@ -0,0 +1,10 @@
+-- Method 2 - use table-valued parameters
+-- create a table type
+CREATE TYPE [mtc_admin].[CheckTableType] AS TABLE (
```
Comment (GuyHarwood) [inline admin/data/sql/migrations/147.do.create-check-table-type.sql]: camel casing for table name please

File: admin/data/sql/migrations/148.do.create-stored-procedure-to-create-checks.sql [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      -- Use RAISERROR inside the CATCH block to return
+      -- error information about the original error that
+      -- caused execution to jump to the CATCH block.
+      RAISERROR (@ErrorMessage, -- Message text.
+        @ErrorSeverity, -- Severity.
+        @ErrorState -- State.
+      );
+  END CATCH
+
+  GO
```
Comment (GuyHarwood) [inline admin/data/sql/migrations/148.do.create-stored-procedure-to-create-checks.sql]: tidy 👌

File: admin/data/sql/migrations/148.do.create-stored-procedure-to-create-checks.sql [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    -- Get the check.id we just inserted
+    SET @checkId = SCOPE_IDENTITY();
+
+    -- Assign a pin to the check
+    INSERT INTO [mtc_admin].[checkPin] (school_id, check_id, pin_id)
+    VALUES (@schoolId, @checkId, (SELECT TOP 1 p.id
+                                  FROM [mtc_admin].[pin] p
+                                         LEFT JOIN [mtc_admin].[checkPin] cp ON (p.id = cp.pin_id)
+                                  WHERE cp.pin_id IS NULL
+                                    AND (cp.school_id IS NULL OR cp.school_id <> 1)));
```
Comment (GuyHarwood) [inline admin/data/sql/migrations/148.do.create-stored-procedure-to-create-checks.sql]: i don't believe we should hardcode this school_id of `1`, as discussed

Comment (jon-shipley) [reply admin/data/sql/migrations/148.do.create-stored-procedure-to-create-checks.sql]: Fixed

File: admin/data/sql/migrations/144.do.create-pin-table.sql (hunk 1 above)
Comment (jon-shipley) [reply admin/data/sql/migrations/144.do.create-pin-table.sql]: done

File: admin/data/sql/migrations/147.do.create-check-table-type.sql (hunk 2 above)
Comment (jon-shipley) [reply admin/data/sql/migrations/147.do.create-check-table-type.sql]: done

File: start.sh [hunk 5]
```diff
@@ -11,8 +11,12 @@ docker-compose up -d
 echo -n "Stabilising MS SQL Server"
 for i in $(seq 1 8); do
     sleep 1
+    if docker exec -it mtc_mssql /opt/mssql-tools/bin/sqlcmd -U SA -P 'Mtc-D3v.5ql_S3rv3r' -Q 'SELECT "database_is_up"' | grep -q database_is_up; then
```
Comment (GuyHarwood) [inline start.sh]: 👌

File: admin/data/sql/migrations/147.do.create-check-table-type.sql [hunk 6]
```diff
@@ -0,0 +1,10 @@
+-- Method 2 - use table-valued parameters
+-- create a table type
+CREATE TYPE [mtc_admin].[checkTableType] AS TABLE (
```
Comment (GeorgeChatzigiannis) [inline admin/data/sql/migrations/147.do.create-check-table-type.sql]: @jon-shipley shouldn't `pupil_id`, `checkForm_id `, `checkWindow_id `and `school_id` be `not null`?
Also `isLiveCheck` should require a default?

Comment (jon-shipley) [reply admin/data/sql/migrations/147.do.create-check-table-type.sql]: @GeorgeChatzigiannis they could be not null.  As it stands though if one of them is null, one of the inserts will fail and the transaction will roll back.

File: functions/package.json [hunk 7]
```diff
@@ -3,7 +3,8 @@
   "scripts": {
     "start": "func host start",
     "test": "jasmine */*.spec.js",
-    "lint": "standard"
+    "lint": "standard",
+    "pack": "funcpack pack ."
```
Comment (jon-shipley) [inline functions/package.json]: @GuyHarwood - can you confirm this?  You deleted this line yesterday, but it seems to be needed by the Validate containers job.

File: admin/data/sql/migrations/147.do.create-check-table-type.sql (hunk 6 above)
Comment (GeorgeChatzigiannis) [reply admin/data/sql/migrations/147.do.create-check-table-type.sql]: @jon-shipley currently they are nullable though? This means the transaction will succeed for any value that is null. Am I missing something?

File: functions/package.json (hunk 7 above)
Comment (GuyHarwood) [reply functions/package.json]: @jon-shipley 
`funcpack` is now deprecated in favour of the new run from package/zip feature.

the lint script should not be removed, that was likely done by accident

Comment (jon-shipley) [reply functions/package.json]: @GuyHarwood - happy to remove it when the "Validate containers &  MIgrations" job no longer requires it.  See the build failure on PR #864 

Merging is being blocked by this.

**Pull request merged**